File: problems/1301-1400/1352_product-of-the-last-k-numbers/solution.cpp
#include <vector>

class ProductOfNumbers {
  public:
    ProductOfNumbers() { prefix.push_back(1); }

    void add(int num) {
        if (num == 0) {
            prefix.resize(1);
            return;
        }
        prefix.push_back(prefix.back() * num);
    }

    int getProduct(int k) {
        if (k >= (int)prefix.size()) {
            return 0;
        }
        return (int)(prefix.back() / prefix[prefix.size() - 1 - k]);
    }

  private:
    std::vector<long long> prefix;
};

File: problems/1301-1400/1352_product-of-the-last-k-numbers/solution.rs
pub struct ProductOfNumbers {
    prefix: Vec<i64>,
}

impl ProductOfNumbers {
    pub fn new() -> Self {
        ProductOfNumbers { prefix: vec![1] }
    }

    pub fn add(&mut self, num: i32) {
        if num == 0 {
            self.prefix.truncate(1);
            return;
        }
        let next = *self.prefix.last().unwrap() * num as i64;
        self.prefix.push(next);
    }

    pub fn getProduct(&mut self, k: i32) -> i32 {
        let size = self.prefix.len();
        if k as usize >= size {
            return 0;
        }
        (self.prefix[size - 1] / self.prefix[size - 1 - k as usize]) as i32
    }
}
